[Order creation] Update coupons display order details

malinajirka opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12619 #12848
Ref: pe5pgL-5G4-p2#comment-4599

This PR updates the how coupons are displayed during Order creation to match the iOS behavior. 

### Changes:

- Updated coupons behavior in the order creation to match the behavior of shipping lines.
- Removed the ability to remove coupons from the totals bottom sheet.


### Notes:
I planned to add unit tests for the added code. However, I haven't figured out how to test the code in the current state without significant refactoring. It felt like the unit tests would be tightly coupled with the inner workings of the VM, which felt off. In any case, if you have some ideas how to write reasonable tests or if you believe it's worth spending more time on, I will be happy to look into it more.

P.S. I also noticed shipping lines are editable at all times which is a bug [I reported here](https://github.com/woocommerce/woocommerce-android/issues/12869).

| Before | After |
|--------|--------|
| ![Screenshot_1730470897](https://github.com/user-attachments/assets/35861b4a-cd82-4392-a2e4-b2e6125ddbb0) | ![Screenshot_1730470794](https://github.com/user-attachments/assets/78ec5d85-15d4-4bd2-a6c4-749446066d74) | 

## Testing

I tested the following flows on my phone and tablet.

**Case: Add coupon to order success:**
* Create a new order, observe that the `+ Add Coupon` row is disabled.
* Add a product to the order, observe that the `+ Add Coupon` row is now enabled.
* Tap on `+ Add Coupon` -> Apply a eligible coupon to that product -> Observe how it appears in the Order, with a "trash bin" button.
* Expand the bottom drawer, observe how the coupon appears in the order details but is not editable from there.

**Case: Delete coupon from order:**
* Tap on the trash icon and observe how the coupon is removed from the order.

**Case: Add coupon to order error**
* Attempt to add a non-eligible coupon to the order (for example: a coupon that is only applicable to product category A, but the order has a product category B)
* Observe how the coupon is not added to the order, and a error notice appears at the bottom

**Case: Editing an existing order**
* Open an order that has coupons applied and is not paid
* Tap on `Edit`
* Observe the coupon rows are editable normally.


**Case: Editing an existing PAID order**
* Open an order that has coupons applied and is paid
* Tap on `Edit`
* Observe the coupon rows are disabled.

I also tested adding and removing shipping lines in combination with coupons with gift card extension installed/uninstalled.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

malinajirka committed:
Add compose implementation of coupons section

malinajirka committed:
Bind model to UI for the coupons section

malinajirka committed:
Make changes to coupons list observable

malinajirka committed:
Disable coupons section when fetching order or in editing mode

malinajirka committed:
Fix UI glitches when adding shipping and coupons

malinajirka committed:
Remove coupon editing from totals bottom sheet

malinajirka committed:
Update release notes

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

malinajirka committed:
Fix detekt

malinajirka committed:
Merge branch 'refs/heads/trunk' into issue/12619-adjust-coupons-behavior-in-order-creation

malinajirka committed:
Fix failing Unit Tests

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditFormFragment.kt:
```diff
@@ -738,6 +760,22 @@ class OrderCreateEditFormFragment :
         } else {
             additionalInfoCollectionSection.addShippingButtonGroup.show()
         }
+
```
This code would ideally live in the VM so it could be unit tested, however, refactoring it felt like out of scope and bringing inconsistency to the surrounding code feels worse than not having this tested. I'm open to suggestions though.

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt:
```diff
@@ -403,10 +405,24 @@ class OrderCreateEditViewModel @Inject constructor(
             }
         }.asLiveData()
 
+    private val _couponLinesLiveData = MediatorLiveData(CouponSection(emptyList(), true))
```
I believe that ideally, we'd have only one viewState for the whole VM or separate state for each part. The current code mixes usages of these two approaches, so I wasn't sure which one to follow. Considering shippingLines are using a separate liveData, I opted to keep consistency with the shipping lines which have almost identical behavior to the coupons section. (perhaps, we use it to avoid having a single state for compose views?)

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt:
```diff
@@ -403,10 +405,24 @@ class OrderCreateEditViewModel @Inject constructor(
         }.asLiveData()
 
+    private val _couponLinesLiveData = MediatorLiveData(CouponSection(emptyList(), true))
+    val couponLinesLiveData = _couponLinesLiveData.distinctUntilChanged()
```
❓  I added `distinctUntilChanged` to ensure updates to the viewStateData and orderDraft cause recomposition only if the coupons related state changes. Having said that, I never dove into compose and I'm not sure if it's necessary as compose might not trigger recomposition if the actual values don't change, even if I didn't use `distinctUntilChanged` - wdyt?

samiuelson replied:
I think `distinctUntilChanged()` helps avoid unnecessary recomposition, because the `CouponSection` model is [unstable](https://developer.android.com/develop/ui/compose/performance/stability#summary) (it contains `List`). As a result, Compose can't know whether `couponLines` value has changed between recompositions.

malinajirka replied:
Nice, thanks for sharing TIL!

malinajirka committed:
Add unit tests for listed coupons lines

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/243">21.1</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

malinajirka committed:
Merge branch 'refs/heads/trunk' into issue/12619-adjust-coupons-behavior-in-order-creation

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt:
```diff
@@ -1803,7 +1819,7 @@ class OrderCreateEditViewModel @Inject constructor(
     }
 
-    private fun onCouponRemoved(couponCode: String) {
+    fun onCouponRemoved(couponCode: String) {
```
💡 N.p. While we're here, maybe we could improve the function name, e.g.:
```suggestion
    fun onRemoveCoupon(couponCode: String) {
```
Feel free to ignore.

malinajirka replied:
Thanks for the suggestion @samiuelson , I renamed both methods to `removeCoupon` and `addCoupon` since in my experience `on` prefix is used for callbacks, but these are not callbacks but methods which actual perform the addition/removal. Let me know in case you'd find this change confusing.

malinajirka replied:
I ended up renaming also the `onAdd` and `onRemove` lambdas as well as related methods to shipping lines to bring consistency between the two.

samiuelson replied:
Nice!

malinajirka committed:
Rename methods names for better clarity

malinajirka committed:
Rename methods names for better clarity

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/coupon/CouponLineFormSection.kt:
```diff
@@ -0,0 +1,152 @@
+                        tint = MaterialTheme.colors.primary
+                    )
+                }
+
```
💡 A minor optimization possibilities: 
* `State<CouponSection>` could be passed to `CouponLineFormSection` Composable as an argument and its `.value` could be read inline in the Composable. That would allow recomposing parts of the Composable (outer, static composables Theme/Card/Column/Row/Text would not be recomposed) instead of the whole.
* instead of `forEachIndexed` [lazy column](https://developer.android.com/develop/ui/compose/lists#item-keys) with key equal to coupon code/id could optimize list rendering.

These are just theoretical considerations/thoughts. I'm not even sure if they are worth applying because the coupons will not be used in big numbers in the order creation.

malinajirka replied:
Thanks for sharing this! Glad I'm learning some compose tricks and gotchas.

I decided not to change this for two reasons:
- The performance benefit is likely not worth the time we'd spend on retesting.
- The current code matches shipping lines and updating both in this PR feels off.

Having said that, if you'd prefer to make this optimization, please create a ticket and I'll work on both coupons and shipping lines in single new PR.

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/coupon/CouponLineFormSection.kt:
```diff
@@ -0,0 +1,152 @@
+        Spacer(modifier = Modifier.weight(1f))
+        Icon(
+            imageVector = Icons.Outlined.DeleteOutline,
+            contentDescription = null,
```
💡 Content descriptions would be nice in the icon and coupon text above.

malinajirka replied:
Good call, fixed in fce8dda3c737775416b11730e41e845569b7b630.

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/coupon/CouponLineFormSection.kt:
```diff
@@ -0,0 +1,152 @@
+}
+
+@Preview
+@Preview(name = "Dark mode", uiMode = Configuration.UI_MODE_NIGHT_YES)
```
```suggestion
@PreviewLightDark
```
💡 Can be used to generate two previews – dark and light

malinajirka replied:
TIL, thanks!

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/coupon/CouponLineFormSection.kt:
```diff
@@ -0,0 +1,152 @@
+            )
+            .padding(dimensionResource(id = R.dimen.major_100))
+    ) {
+        Column(
+            modifier = Modifier
+                .weight(2f)
+                .align(Alignment.CenterVertically)
+        ) {
+            Text(
+                text = couponLine.code,
+                style = TextStyle(
+                    fontWeight = FontWeight.Normal,
+                    fontSize = 18.sp
+                ),
+                color = colorResource(id = R.color.color_on_surface),
+                maxLines = 1,
+                overflow = TextOverflow.Ellipsis
+            )
+        }
+        Spacer(modifier = Modifier.weight(1f))
+        Icon(
+            imageVector = Icons.Outlined.DeleteOutline,
+            contentDescription = null,
+            modifier = Modifier
+                .align(Alignment.CenterVertically)
+                .padding(start = 16.dp)
+        )
+    }
```
💡 I think it would work better simplified—without the `Column` element:
```suggestion
        Text(
            modifier = Modifier.weight(1f),
            text = couponLine.code,
            style = TextStyle(
                fontWeight = FontWeight.Normal,
                fontSize = 18.sp
            ),
            color = colorResource(id = R.color.color_on_surface),
            maxLines = 1,
            overflow = TextOverflow.Ellipsis
        )
        Icon(
            imageVector = Icons.Outlined.DeleteOutline,
            contentDescription = null,
            modifier = Modifier
                .align(Alignment.CenterVertically)
                .padding(start = 16.dp)
        )
```
As a side effect, long coupon codes would look better:
|before|after|
|---|---|
|<img width="425" alt="Screenshot 2024-11-07 at 10 59 59" src="https://github.com/user-attachments/assets/9e642e45-2ce6-43af-9e64-f02aa274fa28">|<img width="422" alt="Screenshot 2024-11-07 at 10 58 32" src="https://github.com/user-attachments/assets/fc498c22-84bb-4e31-9356-601b7f7055e8">|

malinajirka replied:
Thanks, good call! I copy-pasted the shipping lines, but I should have tested it more thoroughly.

malinajirka replied:
Fixed in 825b465632f8e8beb66a2ff3bc2927a2f1a7b0f3

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/shipping/ShippingLineFormSection.kt:
```diff
@@ -42,8 +42,8 @@ import java.math.BigDecimal
-    onAdd: () -> Unit,
-    onEdit: (id: Long) -> Unit,
+    onAddClicked: () -> Unit,
+    onEditClicked: (id: Long) -> Unit,
```
❤️

samiuelson reviewed (commented):
Great work, @malinajirka! I added small improvement ideas. The most important seems to be the one related to long coupon codes' appearance. Feel free to ignore others.

One more thing that caught my attention is the singular form of the "Coupon" label in the bottom sheet despite multiple coupons applied (isn't it slightly confusing?). Maybe worth considering reporting an issue: 
<img width=250 src=https://github.com/user-attachments/assets/6366f751-3331-45f7-94df-b5188cf6b13a/>

malinajirka committed:
Fix appearance of long coupons codes

Co-authored-by: Samuel Urbanowicz <samiuelson@gmail.com>

malinajirka committed:
Display both dark and light mode previews for coupons section

Co-authored-by: Samuel Urbanowicz <samiuelson@gmail.com>

malinajirka committed:
Add content description for remove coupon btn

malinajirka committed:
Update preview

malinajirka committed:
Update coupons line to plural

malinajirka committed:
Remove unused imports

malinajirka commented:
Thanks so much for the thorough review @samiuelson ! I have implemented all the changes except of [the performance optimization](https://github.com/woocommerce/woocommerce-android/pull/12855#discussion_r1832373175). 

> One more thing that caught my attention is the singular form of the "Coupon" label in the bottom sheet despite multiple coupons applied (isn't it slightly confusing?). Maybe worth considering reporting an issue:

I noticed this as well but then forgot about it :X. I updated it in 18e079db3fe4f946ba006148e351589c180868c3. We use plural for products no matter whether there is one or multiple, so the change brings it to consistency.

samiuelson reviewed (approved):
LGTM! :shipit:
